src/core/utils/internal_heap.hpp

```diff
@@ -0,0 +1,354 @@
+     *
+     */
+    bool IsLeftFree(void) const { return GetLeftNext() != 0; }
+
+    /**
+     * This method returns whether the current block is a free block.
+     *
+     * @retval  true    The block is free.
+     * @retval  false   The block is not free.
+     *
+     */
+    bool IsFree(void) const { return mSize != kGuardBlockSize && GetNext() != 0; }
+
+private:
+    enum
+    {
+        kGuardBlockSize = 0xffff, ///< Size value of the guard block.
+    };
+
+    uint16_t mSize; ///< Number of bytes in mMemory.
+
+    /**
+     * Memory for user, with size of *mNext* to ensure size of this
+     * structure is equal to size of block metadata, i.e. sizeof(mSize) + sizeof(mNext)
+     *
+     */
+    uint8_t mMemory[sizeof(uint16_t)];
+};
+
+/**
+ * This class defines functionality to manipulate heap.
+ *
+ * This implementation is currently for mbedTLS.
+ *
+ * The memory is divided into blocks. The whole picture is as follows:
+ *
+ *     +--------------------------------------------------------------------------+
+ *     |    unused      |    super   | block 1 | block 2 | ... | block n | guard  |
+ *     +----------------+------------+---------+---------+-----+---------+--------+
+ *     | kAlignSize - 2 | kAlignSize | 4 + s1  | 4 + s2  | ... | 4 + s4  |   2    |
+ *     +--------------------------------------------------------------------------+
+ *
+ */
+class InternalHeap
+{
+public:
+    /**
+     * This function initializes a memory heap.
+     *
+     */
+    void Init(void);
+
+    /**
+     * This method allocates at least @p aCount * @aSize bytes memory and initialize to zero.
+     *
+     * @param[in]   aCount  Number of allocate units.
+     * @param[in]   aSize   Unit size in bytes.
+     *
+     * @returns A pointer to the allocated memory.
+     *
+     * @retval  NULL    Indicates not enough memory.
+     *
+     */
+    void *CAlloc(size_t aCount, size_t aSize);
+
+    /**
+     * This method free memory pointed by @p aPointer.
+     *
+     * @param[in]   aPointer    A pointer to the memory to free.
+     *
+     */
+    void Free(void *aPointer);
+
+    /**
+     * This method returns whether the heap is clean.
+     *
+     */
+    bool IsClean(void) const
+    {
+        InternalHeap &self  = *const_cast<InternalHeap *>(this);
+        const Block & super = self.BlockSuper();
+        const Block & first = self.BlockRight(super);
+        return super.GetNext() == self.BlockOffset(first) && first.GetSize() == kFirstBlockSize;
+    }
+
+    /**
+     * This method returns the capacity of this heap.
+     *
+     */
+    size_t GetCapacity(void) const { return kFirstBlockSize; }
+
+    /**
+     * This method returns free space of this heap.
+     */
+    size_t GetFreeSize(void) const { return mMemory.mFreeSize; }
+
+private:
+    enum
+    {
+#if OPENTHREAD_ENABLE_DTLS
+        kMemorySize = OPENTHREAD_CONFIG_HEAP_SIZE, ///< Size of memory buffer (bytes).
```

Comment:
What about to change the description (and potentially name - to include `INTERNAL` word) to `OPENTHREAD_CONFIG_HEAP_SIZE` and `OPENTHREAD_CONFIG_HEAP_SIZE_NO_DTLS`. Just to make it clear that those configs are used when `OPENTHREAD_CONFIG_EXTERNAL_HEAP` is disabled.